third_party/amd/lib/TritonAMDGPUToLLVM/OptimizeLDSUsage.cpp

```diff
@@ -0,0 +1,305 @@
+// layout uses across y dimension.
+//
+// clang-format off
+//
+// LDS usage of this op is roughly calculated as:
+// LDS_USAGE = getShapePerCTA(mfma_layout)[0] * getShapePerCTA(blocked_layout)[1] * sizeof(data_type)
+// LDS_USAGE = warpsPerCTA(mfma_layout)[0] * warpsPerCta(blocked_layout)[1] * C,
+// where C = 32 * sizePerWarp(blocked_layout)[1] * threadsPerWarp(blocked_layout)[1] * sizeof(data_type)
+//
+// clang-format on
+//
+// When LDS_USAGE exceeds the size of LDS, try to lower LDS usage by
+// decomposing cvt(mfma->blocked) op into 2 conversions: cvt(mfma->mfma_tmp)
+// and cvt(mfma_tmp->blocked), where mfma_tmp has WarpsPerCta attribute that
+// minimizes uses of LDS for these conversions.
+static void tryMinimizeLDS(triton::gpu::ConvertLayoutOp cvtOp) {
+  OpBuilder builder(cvtOp);
+
+  auto srcType = cvtOp.getSrc().getType();
+  auto dstType = cvtOp.getType();
+
+  auto srcEnc = srcType.getEncoding();
+  auto dstEnc = dstType.getEncoding();
+
+  auto currLDSUsage = getCvtOpLDSUsage(cvtOp);
+  if (currLDSUsage <= LDSSize) {
+    return;
+  }
+
+  auto ctx = srcEnc.getContext();
+  auto rank = srcType.getShape().size();
+  unsigned numWarps = triton::gpu::getNumWarpsPerCTA(srcEnc);
+  auto warpSize = triton::gpu::getWarpSize(srcEnc);
+
+  triton::gpu::ConvertLayoutOp tmpCvt;
+  triton::gpu::ConvertLayoutOp newEpilogueCvt;
+
+  // Find all possible shapes of WarpsPerCTA by finding all possible
+  // factorizations of numWarps. Pick shape for which both conversions in
+  // decomposition use LDS less than LDSSize and for which sum of LDS usage
+  // is minimal. If no such shape exists, do not decompose.
+  auto factorizedNumWarps = factorizePowerOf2(numWarps);
+  // Create a list of temporary layouts
+  SmallVector<Attribute> tmpLayouts;
+  for (int i = 0; i < factorizedNumWarps.size(); i++) {
+    auto warpsPerCTA = factorizedNumWarps[i];
+    tmpLayouts.push_back(createTmpLayout(srcEnc, warpsPerCTA));
+    tmpLayouts.push_back(createTmpLayout(dstEnc, warpsPerCTA));
+    SmallVector<unsigned> elemsPerThread(rank, 1);
+    SmallVector<unsigned> threadsPerWarp(rank, 1);
+    threadsPerWarp[rank - 1] = warpSize / 8;
+    threadsPerWarp[rank - 2] = warpSize / threadsPerWarp[rank - 1];
+    auto order = triton::gpu::getOrder(srcEnc);
+    auto layoutCTA = triton::gpu::getCTALayout(srcEnc);
+    auto fallbackLayout = triton::gpu::BlockedEncodingAttr::get(
+        ctx, elemsPerThread, threadsPerWarp, warpsPerCTA, order, layoutCTA);
+    tmpLayouts.push_back(fallbackLayout);
+  }
+
+  unsigned minLDSUsage = 2 * LDSSize;
+  int minIdx = -1;
+  for (int i = 0; i < tmpLayouts.size(); i++) {
+    auto tmpLayout = tmpLayouts[i];
+    std::tie(tmpCvt, newEpilogueCvt) =
+        createNewConvertOps(builder, cvtOp, tmpLayout);
+
+    int tmpCvtLDS = getCvtOpLDSUsage(tmpCvt);
+    int newCvtLDS = getCvtOpLDSUsage(newEpilogueCvt);
+    if (tmpCvtLDS <= LDSSize && newCvtLDS <= LDSSize) {
+      int LDSUsage = std::max(tmpCvtLDS, newCvtLDS);
```

Comment:
@oplavsic 
I've changed this part of the algorithm: https://github.com/openai/triton/pull/3730/files#diff-0d63e5cd9cf58151489fd9a5206b43a0902939004e58f3a7ec5258fa7d473267L227

Was it crucial?